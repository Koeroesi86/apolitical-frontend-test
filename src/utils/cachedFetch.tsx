import axios from 'axios';
import cacheProvider from './cacheProvider';

export default async (url) => {
  const cached = await cacheProvider.get(url);
  if (cached !== undefined) return cached;

  const data = await axios({url, method: 'GET'}).then(res => res.data);
  await cacheProvider.set(url, data);

  return data;
}