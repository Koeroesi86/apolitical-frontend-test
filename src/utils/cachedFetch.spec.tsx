import moxios from 'moxios';
import cachedFetch from './cachedFetch';

describe('cachedFetch', () => {
  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('should call axios only once', async () => {
    const url = 'https://example.com';
    moxios.stubRequest(url, {
      status: 200,
      response: {},
    });

    await cachedFetch(url);
    await cachedFetch(url);

    expect(moxios.requests.count()).toEqual(1);
    expect(moxios.requests.mostRecent()).toMatchSnapshot();
  });
});
