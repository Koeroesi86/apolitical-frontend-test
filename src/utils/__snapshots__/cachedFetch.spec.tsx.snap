// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cachedFetch should call axios only once 1`] = `
Request {
  "config": Object {
    "adapter": [Function],
    "data": undefined,
    "headers": Object {
      "Accept": "application/json, text/plain, */*",
    },
    "maxBodyLength": -1,
    "maxContentLength": -1,
    "method": "get",
    "timeout": 0,
    "transformRequest": Array [
      [Function],
    ],
    "transformResponse": Array [
      [Function],
    ],
    "url": "https://example.com",
    "validateStatus": [Function],
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "headers": Object {
    "Accept": "application/json, text/plain, */*",
  },
  "reject": [Function],
  "resolve": [Function],
  "responseType": undefined,
  "timeout": 0,
  "url": "https://example.com",
  "withCredentials": false,
}
`;
